src/refdb_reftable.c

```diff
@@ -0,0 +1,514 @@
+			break;
+		case GIT_REFERENCE_DIRECT:
+			r.value = (unsigned char *) git_reference_target(ref)->id;
+			/* TODO: peel to commit */
+			r.target_value = (unsigned char *) git_reference_target_peel(ref)->id;
+			break;
+		default:
+			return -1;
+	}
+
+	if ((error = reftable_log_fill(&l, who, old_id, git_reference_target_peel(ref),
+				       git_reference_name(ref), message)) < 0 ||
+	    (error = stack_add(backend->stack, reftable_writer, &arg)) < 0)
+		goto out;
+
+out:
+	return error;
+}
+
+static int reftable_delete(
+	git_refdb_backend *backend,
+	const char *ref_name,
+	const git_oid *old_id,
+	const char *old_target)
+{
+	struct ref_record ref = { 0 };
+	struct log_record log = { 0 };
+	int error;
+
+	GIT_UNUSED(old_target);
+
+	ref.ref_name = (char *)ref_name;
+
+	if ((error = reftable_log_fill(&log, NULL, old_id, NULL, ref_name, NULL)) < 0 ||
+	    (error = reftable_write_records(backend, &ref, 1, &log, 1)) < 0)
+		goto out;
+
+out:
+	return error;
+}
+
+static int reftable_rename(
+	git_reference **out,
+	git_refdb_backend *_backend,
+	const char *old_name,
+	const char *new_name,
+	int force,
+	const git_signature *who,
+	const char *message)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record refs[2] = {{0}, {0}}, old = {0};
+	struct log_record logs[2] = {{0}, {0}};
+	git_oid oid;
+	int error;
+
+	GIT_UNUSED(force);
+
+	if ((error = stack_read_ref(backend->stack, old_name, &old)) != 0) {
+		if (error > 0)
+			error = GIT_ENOTFOUND;
+		goto out;
+	}
+
+	if ((error = git_oid_fromraw(&oid, old.target_value)) < 0)
+		goto out;
+
+	refs[0].ref_name = (char *) old_name;
+	refs[1].ref_name = (char *) new_name;
+	refs[1].value = old.value;
+	refs[1].target_value = old.target_value;
+	refs[1].target = old.target;
+
+	if ((error = reftable_log_fill(&logs[0], who, &oid, NULL, old_name, message)) < 0 ||
+	    (error = reftable_log_fill(&logs[1], who, NULL, &oid, new_name, message)) < 0 ||
+	    (error = reftable_write_records(_backend, refs, 2, logs, 2)) < 0)
+		goto out;
+
+	if ((error = reftable_reference_from_record(out, &refs[1])) < 0)
+		goto out;
+
+out:
+	ref_record_clear(&old);
+	return error;
+}
+
+static void reftable_free(git_refdb_backend *_backend)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	stack_destroy(backend->stack);
+	git__free(backend);
+}
+
+static int reftable_reflog__has_log(git_refdb_backend *_backend, const char *refname)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record record;
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, refname, &record)) < 0)
+		return error;
+	if (error > 0)
+		return 0;
+
+	ref_record_clear(&record);
+	return 1;
+}
+
+static int reftable_reflog__ensure_log(git_refdb_backend *_backend, const char *name)
+{
+	/* TODO: doesn't make any sense in the context of reftable? */
```

Comment:
I don't think this makes any sense in the context of reftable, but I'd say it is our issue and not reftable's.

Reply:
ack.